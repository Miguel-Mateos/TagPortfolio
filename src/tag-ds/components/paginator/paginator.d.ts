import React from 'react';
export interface IPaginator {
    /**
     * Number of pages existing
     */
    pages: number;
    /**
     * First page to be shown
     */
    defaultPage?: number;
    /**
     * Selector for the collapsed Variant
     */
    collapsed?: boolean;
    className?: string;
    /**
     * For truncating the default Paginator
     */
    truncate?: 'left' | 'right' | 'all';
    /**
     * In case of having collapsed variant,
     * options for inserting the jump element on different positions
     */
    collapsedOptions?: {
        jump?: 'right' | 'left' | 'all';
    };
    disabled?: boolean;
    onChangePage?: (page: number) => void;
    [key: string]: any;
}
export declare const Paginator: React.FC<IPaginator>;
