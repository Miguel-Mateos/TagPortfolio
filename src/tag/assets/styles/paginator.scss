%paginator-page {
  border-radius: 50%;
  border-color: transparent;
  @extend %button-secondary;
  min-width: auto;
  color: var(--neutral800);
  height: calc(3 * var(--unit));
  width: calc(3 * var(--unit));
  padding: 0;
  display: flex;
  justify-content: center;
  align-items: center;
}

.tag-ds {
  .paginator,
  &.paginator {
    @include font(small);
    display: flex;
    gap: var(--unit);
  }

  .paginator-page_active {
    @extend %paginator-page;
    background-color: var(--accent500);

    &:focus {
      border: calc(var(--unit) / 8) solid var(--accent500);
    }
    &.disabled {
      background-color: var(--neutral200);
    }

    &:disabled {
      border-color: transparent;
      background-color: var(--neutral200);
      color: var(--neutral500);
    }
  }

  .paginator-truncate {
    margin: 0;
  }

  .paginator-page {
    @extend %paginator-page;

    &:disabled {
      border-color: transparent;
      color: var(--neutral500);
    }
  }

  .paginator-collapsed {
    display: flex;
    gap: var(--unit);
    align-items: center;
    color: var(--neutral800);
  }

  .paginator-collapsed-dropdown {
    width: calc(11 * var(--unit));
    height: calc(5 * var(--unit));
  }

  .paginator-icon {
    user-select: none;
    display: flex;
    justify-content: center;
    align-items: center;
    height: calc(3 * var(--unit));
    width: calc(3 * var(--unit));
    padding: 0;
    @extend %button-secondary;
    border-color: transparent;
    min-width: auto;
    & > span {
      user-select: none;
      font-size: calc(2 * var(--unit));
      color: var(--neutral800);
    }
    &.disabled {
      & > span {
        color: var(--neutral500);
      }
    }

    &:disabled {
      & > span {
        color: var(--neutral500);
      }
    }
  }
}
