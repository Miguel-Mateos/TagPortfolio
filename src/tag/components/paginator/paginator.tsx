import React, {useState} from 'react';
import Select, {Option} from '../select/select';

export interface IPaginator {
  /**
   * Number of pages existing
   */
  pages: number;
  /**
   * First page to be shown
   */
  defaultPage?: number;
  /**
   * Selector for the collapsed Variant
   */
  collapsed?: boolean;
  className?: string;
  /**
   * For truncating the default Paginator
   */
  truncate?: 'left' | 'right' | 'all';
  /**
   * In case of having collapsed variant,
   * options for inserting the jump element on different positions
   */
  collapsedOptions?: {
    jump?: 'right' | 'left' | 'all';
  };
  disabled?: boolean;
  onChangePage?: (page: number) => void;
  [key: string]: any;
}

export const Paginator: React.FC<IPaginator> = ({
  pages,
  defaultPage,
  collapsed = false,
  collapsedOptions,
  disabled = false,
  truncate,
  className,
  onChangePage,
  ...rest
}) => {
  const [current, setCurrent] = useState<number>(defaultPage || 1);
  const arr = Array.from({length: pages}, (_, idx) => idx + 1);
  const setTruncation = () => {
    if (truncate === 'left') return arr.slice(1, 6);
    if (truncate === 'right') return arr.slice(0, 5);
    return arr.slice(0, 5);
  };

  const [chunk, setChunk] = useState<number[]>(arr.length > 6 ? setTruncation : arr);

  const handleChunkIncrement = () => {
    if (truncate === 'right') {
      const newArr = arr.slice(chunk[chunk.length - 2], chunk[chunk.length - 2] + 5);
      if (newArr[newArr.length - 1] === arr[arr.length - 1])
        setChunk(arr.slice(chunk[chunk.length - 2], arr[arr.length - 2]));
    }
    setChunk(arr.slice(chunk[chunk.length - 2], chunk[chunk.length - 2] + 5));
  };

  const handleChunkDecrement = () => {
    if (chunk[0] !== arr[1]) setChunk(arr.slice(chunk[0] - 5, chunk[0]));
  };

  const onChange = (page: number) => {
    setCurrent(page);
    if (page === arr[0]) {
      setChunk(arr.slice(0, 5));
    } else if (page === arr[arr.length - 1]) {
      setChunk(arr.slice(arr.length - 5, arr.length));
    }
    if (typeof onChangePage === 'function') onChangePage(page);
  };

  if (collapsed) {
    return (
      <div
        className={`paginator-collapsed ${className || ''}`}
        data-testid={rest && rest['data-testid'] ? rest['data-testid'] : undefined}
      >
        {(collapsedOptions?.jump === 'left' || collapsedOptions?.jump === 'all') && (
          <button
            disabled={disabled}
            className={`paginator-icon ${current === arr[0] ? 'disabled' : ''}`}
            data-testid={rest && rest['data-testid'] ? `${rest['data-testid']}-button-jump-previous` : ''}
            onClick={() => onChange(arr[0])}
          >
            <span className="material-icons">first_page</span>
          </button>
        )}
        {arr.length !== 0 && (
          <button
            disabled={disabled}
            className={`paginator-icon ${current === arr[0] ? 'disabled' : ''}`}
            data-testid={rest && rest['data-testid'] ? `${rest['data-testid']}-button-previous` : ''}
            onClick={() => current !== arr[0] && onChange(current - 1)}
          >
            <span className="material-icons">navigate_before</span>
          </button>
        )}
        <Select
          disabled={disabled}
          name="paginator-selector"
          placeholder={defaultPage?.toString() || '1'}
          className="paginator-collapsed-dropdown"
          value={current}
          data-testid={rest && rest['data-testid'] ? `${rest['data-testid']}-dropdown` : ''}
          onChange={onChange}
        >
          {arr &&
            arr.length > 0 &&
            arr.map((num) => (
              <Option
                key={num + 'select-item'}
                value={num}
                label={num.toString()}
                data-testid={rest && rest['data-testid'] ? `${rest['data-testid']}-dropdown-item-${num}` : ''}
              >
                {num}
              </Option>
            ))}
        </Select>
        <span>of {arr.length} pages</span>
        {arr.length !== 0 && (
          <button
            disabled={disabled}
            className={`paginator-icon ${current === arr[arr.length - 1] ? 'disabled' : ''}`}
            data-testid={rest && rest['data-testid'] ? `${rest['data-testid']}-button-next` : ''}
            onClick={() => current !== arr[arr.length - 1] && onChange(current + 1)}
          >
            <span className="material-icons">navigate_next</span>
          </button>
        )}
        {(collapsedOptions?.jump === 'right' || collapsedOptions?.jump === 'all') && (
          <button
            disabled={disabled}
            className={`paginator-icon ${current === arr[arr.length - 1] ? 'disabled' : ''}`}
            data-testid={rest && rest['data-testid'] ? `${rest['data-testid']}-button-jump-next` : ''}
            onClick={() => onChange(arr[arr.length - 1])}
          >
            <span className="material-icons">last_page</span>
          </button>
        )}
      </div>
    );
  }

  return (
    <div
      className={`paginator ${className || ''}`}
      data-testid={rest && rest['data-testid'] ? rest['data-testid'] : undefined}
    >
      {arr.length > 6 && (
        <button
          disabled={disabled}
          className={`paginator-icon ${chunk[0] === arr[1] || chunk[0] === arr[0] ? 'disabled' : ''}`}
          onClick={handleChunkDecrement}
          data-testid={rest && rest['data-testid'] ? `${rest['data-testid']}-button-previous` : ''}
        >
          <span className="material-icons">navigate_before</span>
        </button>
      )}
      {(truncate === 'left' || truncate === 'all') && chunk[0] > arr[3] && (
        <>
          <button
            className={current === arr[0] ? 'paginator-page_active' : 'paginator-page'}
            data-testid={rest && rest['data-testid'] ? `${rest['data-testid']}-truncate-left` : ''}
            onClick={() => onChange(arr[0])}
          >
            {arr[0]}
          </button>
          <span className="paginator-truncate">...</span>
        </>
      )}
      {chunk.map((_num) => (
        <button
          disabled={disabled}
          className={current === _num ? 'paginator-page_active' : 'paginator-page'}
          onClick={() => onChange(_num)}
          key={_num + 'paginator-num-page'}
        >
          {_num}
        </button>
      ))}
      {(truncate === 'right' || truncate === 'all') && chunk[chunk.length - 1] < arr[arr.length - 3] && (
        <>
          <span className="paginator-truncate">...</span>
          <button
            disabled={disabled}
            className={current === arr[arr.length - 1] ? 'paginator-page_active' : 'paginator-page'}
            data-testid={rest && rest['data-testid'] ? `${rest['data-testid']}-truncate-right` : ''}
            onClick={() => onChange(arr[arr.length - 1])}
          >
            {arr[arr.length - 1]}
          </button>
        </>
      )}
      {arr.length > 6 && (
        <button
          disabled={disabled}
          className={`paginator-icon ${chunk[chunk.length - 1] === arr[arr.length - 1] ? 'disabled' : ''}`}
          onClick={handleChunkIncrement}
          data-testid={rest && rest['data-testid'] ? `${rest['data-testid']}-button-next` : ''}
        >
          <span className="material-icons">navigate_next</span>
        </button>
      )}
    </div>
  );
};
