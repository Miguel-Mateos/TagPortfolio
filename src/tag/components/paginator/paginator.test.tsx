import React from 'react';
import {IPaginator, Paginator} from './paginator';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import {fireEvent, logRoles, render, screen} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

const PaginatorTest = (args: IPaginator) => {
  return <Paginator {...args} data-testid="paginator" />;
};

test('renders and match snapshot', () => {
  const component = renderer.create(<PaginatorTest pages={20} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Renders default paginator', () => {
  const {getByTestId} = render(<PaginatorTest pages={20} />);
  const paginator = getByTestId('paginator');
  expect(paginator.querySelector('paginator')).toBeDefined();
});

test('renders collapsed paginator', () => {
  const {getByTestId} = render(<PaginatorTest pages={20} collapsed />);
  const paginator = getByTestId('paginator');
  expect(paginator.querySelector('paginator-collapsed')).toBeDefined();
});

test('incerment page default', () => {
  const {getByText, getByTestId} = render(<PaginatorTest pages={20} />);
  const nextButton = getByTestId('paginator-button-next');
  expect(getByText(1)).toBeDefined();
  userEvent.click(nextButton);
  expect(getByText(5)).toBeDefined();
});

test('incerment page truncate', () => {
  const {getByText, getByTestId} = render(<PaginatorTest pages={20} truncate="right" />);
  const nextButton = getByTestId('paginator-button-next');
  expect(getByText(1)).toBeDefined();
  userEvent.click(nextButton);
  expect(getByText(5)).toBeDefined();
});

test('decrement page default', () => {
  const {getByText, getByTestId} = render(<PaginatorTest pages={20} defaultPage={6} />);
  const nextButton = getByTestId('paginator-button-next');
  const prevButton = getByTestId('paginator-button-previous');
  userEvent.click(nextButton);
  expect(getByText(7)).toBeDefined();
  userEvent.click(prevButton);
  expect(getByText(1)).toBeDefined();
});

test('decrement page truncated', () => {
  const {getByText, getByTestId} = render(<PaginatorTest pages={20} truncate="right" />);
  const nextButton = getByTestId('paginator-button-next');
  const prevButton = getByTestId('paginator-button-previous');
  userEvent.click(nextButton);
  expect(getByText(7)).toBeDefined();
  userEvent.click(prevButton);
  expect(getByText(1)).toBeDefined();
});

test('truncate to start (left)', () => {
  const {getByText, getByTestId} = render(<PaginatorTest pages={20} truncate="left" />);
  const nextChunckBtn = getByTestId('paginator-button-next');
  if (nextChunckBtn) fireEvent.click(nextChunckBtn);
  const trunc = getByTestId('paginator-truncate-left');
  userEvent.click(trunc);
  expect(getByText(1)).toHaveClass('paginator-page_active');
});

test('truncate to finish (right)', () => {
  const {getByText, getByTestId} = render(<PaginatorTest pages={20} truncate="right" />);
  const trunc = getByTestId('paginator-truncate-right');
  userEvent.click(trunc);
  expect(getByText(20)).toHaveClass('paginator-page_active');
});

test('jump option to start', () => {
  const {getByTestId, container} = render(
    <PaginatorTest pages={20} collapsed defaultPage={5} collapsedOptions={{jump: 'all'}} />
  );
  const jump = getByTestId('paginator-button-jump-previous');

  userEvent.click(jump);
  expect(container.getElementsByClassName('input')[0]?.innerHTML).toBe('1');
});

test('jump option to finish', () => {
  const {getByTestId, container} = render(
    <PaginatorTest pages={20} collapsed defaultPage={5} collapsedOptions={{jump: 'all'}} />
  );
  const jump = getByTestId('paginator-button-jump-next');

  userEvent.click(jump);
  expect(container.getElementsByClassName('input')[0]?.innerHTML).toBe('20');
});

test('collapsed arrow navigation left', () => {
  const {getByTestId, container} = render(
    <PaginatorTest pages={20} collapsed defaultPage={5} collapsedOptions={{jump: 'all'}} />
  );
  const previous = getByTestId('paginator-button-previous');
  userEvent.click(previous);
  expect(container.getElementsByClassName('input')[0]?.innerHTML).toBe('4');
});

test('collapsed arrow navigation right', () => {
  const {getByTestId, container} = render(
    <PaginatorTest pages={20} collapsed defaultPage={5} collapsedOptions={{jump: 'all'}} />
  );
  const next = getByTestId('paginator-button-next');
  userEvent.click(next);
  expect(container.getElementsByClassName('input')[0]?.innerHTML).toBe('6');
});

test('select page from dropdown collapsed', () => {
  const {getByTestId} = render(<PaginatorTest pages={20} collapsed defaultPage={5} collapsedOptions={{jump: 'all'}} />);
  const dropdown = getByTestId('paginator-dropdown');
  if (dropdown) userEvent.click(dropdown);
  const DItem = screen.getByTestId('paginator-dropdown-item-4');
  userEvent.click(DItem);
  expect(dropdown.getElementsByClassName('input')[0]?.innerHTML).toBe('4');
});

test('minimum pages render', () => {
  const {queryByText} = render(<PaginatorTest pages={5} />);
  expect(queryByText(/navigate_next/i)).toBeNull();
});

test('disabled paginator', () => {
  const {getByTestId} = render(<PaginatorTest pages={20} disabled />);
  const disabledNumber = getByTestId('paginator-button-next');
  expect(disabledNumber).toHaveAttribute('disabled');
});
