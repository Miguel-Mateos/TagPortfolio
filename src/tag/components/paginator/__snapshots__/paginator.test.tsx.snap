// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders and match snapshot 1`] = `
<div
  className="paginator "
  data-testid="paginator"
>
  <button
    className="paginator-icon disabled"
    data-testid="paginator-button-previous"
    disabled={false}
    onClick={[Function]}
  >
    <span
      className="material-icons"
    >
      navigate_before
    </span>
  </button>
  <button
    className="paginator-page_active"
    disabled={false}
    onClick={[Function]}
  >
    1
  </button>
  <button
    className="paginator-page"
    disabled={false}
    onClick={[Function]}
  >
    2
  </button>
  <button
    className="paginator-page"
    disabled={false}
    onClick={[Function]}
  >
    3
  </button>
  <button
    className="paginator-page"
    disabled={false}
    onClick={[Function]}
  >
    4
  </button>
  <button
    className="paginator-page"
    disabled={false}
    onClick={[Function]}
  >
    5
  </button>
  <button
    className="paginator-icon "
    data-testid="paginator-button-next"
    disabled={false}
    onClick={[Function]}
  >
    <span
      className="material-icons"
    >
      navigate_next
    </span>
  </button>
</div>
`;
